### PR TITLE
Binary policy read fuzzer

### PR DESCRIPTION
**Pull request by cgzones**
Introduce a libfuzz[1] based fuzzer testing the parsing of a binary
policy.

Build the fuzzer in the oss-fuzz script.

[1]: https://llvm.org/docs/LibFuzzer.html

Additionally fix several reported issues.

@evverx @fishilico

### CODE REVIEW

**Comment by evverx** [comment]
CIFuzz failed due to https://github.com/google/clusterfuzz/pull/2471 as far as I can tell. The next push should trigger CIFuzz without that bug I think

File: scripts/oss-fuzz.sh [hunk 1]
```diff
@@ -32,10 +32,10 @@ SANITIZER=${SANITIZER:-address}
 flags="-O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=$SANITIZER -fsanitize=fuzzer-no-link"
 
 export CC=${CC:-clang}
-export CFLAGS=${CFLAGS:-$flags}
+export CFLAGS="${CFLAGS:-$flags} -I$DESTDIR/usr/include -D_GNU_SOURCE -D_FILE_OFFSET_BITS=64 -Wall -Wextra"
```
[the remark this answers is not in the document]
Comment (cgzones) [reply scripts/oss-fuzz.sh]: Dropped.

**Comment by evverx** [comment]
> The next push should trigger CIFuzz without that bug I think

Unfortunately, it seems the latest version of libClusterFuzz hasn't been released yet. Until it's updated on OSS-Fuzz/CIFuzz that issue can be "fixed" temporarily by reverting the commit where `report-unreproducible-crashes` is set to `true`.

File: scripts/oss-fuzz.sh [hunk 2]
```diff
@@ -49,11 +49,24 @@ make -C libsepol clean
...
+## binary policy fuzzer ##
+
+# CFLAGS, CXXFLAGS and LIB_FUZZING_ENGINE have to be split to be accepted by
+# the compiler/linker so they shouldn't be quoted
+# shellcheck disable=SC2086
+$CC $CFLAGS -c -o binpolicy-fuzzer.o libsepol/fuzz/binpolicy-fuzzer.c
+# shellcheck disable=SC2086
+$CXX $CXXFLAGS $LIB_FUZZING_ENGINE binpolicy-fuzzer.o "$DESTDIR/usr/lib/libsepol.a" -o "$OUT/binpolicy-fuzzer"
+
+zip -j "$OUT/secilc-fuzzer_seed_corpus.zip" libsepol/fuzz/policy.bin
```
Comment (evverx) [inline scripts/oss-fuzz.sh:72]: Thanks! I think it should be called `binpolicy-fuzzer_seed_corpus.zip` instead of `secilc-fuzzer_seed_corpus.zip`

**Comment by doverride** [comment]
cgzones ***@***.***> writes:

> @evverx I noticed the step Build Fuzzers is taking more than 3 minutes due to the command
> apt-get install -y bison flex gawk gettext make libaudit-dev libbz2-dev libcap-dev
> libcap-ng-dev libglib2.0-dev libpcre3-dev xmlto
> (https://github.com/google/oss-fuzz/blob/fd15c6d57aa13667af2521daf6167a2d8cd2ecb7/projects/selinux/Dockerfile#L19).
>
> Is there a more suitable base image pre-containing more of the required packages or would
> adding the apt-get option --no-install-recommends reduce the amount of installed packages?

-yy implies --no-install-recommends

xmlto probably pulls in a lot of stuff and afaik you only need that for
building secilc man pages unfortunately that is implied with make
install (but a `make secilc` does not require xmlto). If this is just
about libsepol then I suspect one might be able to exclude a bunch of
those build deps.

>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android. 
>

-- 
gpg --locate-keys ***@***.***
Key fingerprint = FCD2 3660 5D6B 9D27 7FC6  E0FF DA7E 521F 10F6 4098
Dominick Grift

**Comment by cgzones** [comment]
> while I'm at it I can add your email address to https://github.com/google/oss-fuzz/blob/master/projects/selinux/project.yaml if it's fine by you.

I'll ask the maintainers if it's ok for them once this patch-set is about to be merged.

**Comment by evverx** [comment]
@cgzones it seems OSS-Fuzz started reporting bugs found by the fuzzer:
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=42697
```
==462==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000008 (pc 0x00000050a2f0 bp 0x7fff482b9050 sp 0x7fff482b8f40 T0)
--
  | ==462==The signal is caused by a READ memory access.
  | ==462==Hint: address points to the zero page.
  | SCARINESS: 10 (null-deref)
  | #0 0x50a2f0 in build_type_map selinux/libsepol/src/optimize.c:107:33
  | #1 0x50a2f0 in policydb_optimize selinux/libsepol/src/optimize.c:441:13
  | #2 0x4e12f3 in LLVMFuzzerTestOneInput selinux/libsepol/fuzz/binpolicy-fuzzer.c:42:10
  | #3 0x4d725b in main
  | #4 0x7f69776920b2 in __libc_start_main /build/glibc-eX1tMB/glibc-2.31/csu/libc-start.c:308:16
  | #5 0x41d4ed in _start
```

**Pull request closed without merging**